File: backend/src/main/java/com/infocurso/backend/service/AsistenciaService.java
package com.infocurso.backend.service;

import com.infocurso.backend.dto.AsistenciaDTO;
import com.infocurso.backend.dto.EstadisticasAsistenciaDTO;
import com.infocurso.backend.dto.RegistroAsistenciaDTO;
import com.infocurso.backend.entity.Asistencia;
import com.infocurso.backend.entity.Curso;
import com.infocurso.backend.entity.EstadoAsistencia;
import com.infocurso.backend.entity.Usuario;
import com.infocurso.backend.repository.AsistenciaRepository;
import com.infocurso.backend.repository.CursoRepository;
import com.infocurso.backend.repository.UsuarioRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.web.server.ResponseStatusException;

import java.time.LocalDate;
import java.util.List;
import java.util.UUID;

@Service
@RequiredArgsConstructor
public class AsistenciaService {

    private final AsistenciaRepository asistenciaRepository;
    private final CursoRepository cursoRepository;
    private final UsuarioRepository usuarioRepository;

    public List<AsistenciaDTO> getAsistenciaCurso(UUID cursoId) {
        return asistenciaRepository.findByCursoIdOrderByFechaDesc(cursoId)
                .stream()
                .map(AsistenciaDTO::from)
                .toList();
    }

    public List<AsistenciaDTO> getAsistenciaCursoByFecha(UUID cursoId, LocalDate fecha) {
        return asistenciaRepository.findByCursoIdAndFecha(cursoId, fecha)
                .stream()
                .map(AsistenciaDTO::from)
                .toList();
    }

    public List<AsistenciaDTO> getAsistenciaAlumno(UUID alumnoId, UUID cursoId) {
        return asistenciaRepository.findByAlumnoIdAndCursoIdOrderByFechaDesc(alumnoId, cursoId)
                .stream()
                .map(AsistenciaDTO::from)
                .toList();
    }

    public AsistenciaDTO registrarAsistencia(UUID cursoId, RegistroAsistenciaDTO dto) {
        Curso curso = cursoRepository.findById(cursoId)
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, "Curso no encontrado"));

        Usuario alumno = usuarioRepository.findById(dto.alumnoId())
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, "Alumno no encontrado"));

        // Verificar si ya existe un registro para ese día
        Asistencia asistencia = asistenciaRepository
                .findByAlumnoIdAndCursoIdAndFecha(dto.alumnoId(), cursoId, dto.fecha())
                .orElse(null);

        if (asistencia == null) {
            asistencia = Asistencia.builder()
                    .alumno(alumno)
                    .curso(curso)
                    .fecha(dto.fecha())
                    .estado(dto.estado())
                    .observaciones(dto.observaciones())
                    .build();
        } else {
            asistencia.setEstado(dto.estado());
            asistencia.setObservaciones(dto.observaciones());
        }

        asistencia = asistenciaRepository.save(asistencia);
        return AsistenciaDTO.from(asistencia);
    }

    public void registrarAsistenciaMultiple(UUID cursoId, List<RegistroAsistenciaDTO> registros) {
        for (RegistroAsistenciaDTO dto : registros) {
            registrarAsistencia(cursoId, dto);
        }
    }

    public AsistenciaDTO actualizarAsistencia(UUID asistenciaId, EstadoAsistencia estado, String observaciones) {
        Asistencia asistencia = asistenciaRepository.findById(asistenciaId)
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, "Registro no encontrado"));

        asistencia.setEstado(estado);
        asistencia.setObservaciones(observaciones);
        asistencia = asistenciaRepository.save(asistencia);
        return AsistenciaDTO.from(asistencia);
    }

    public void eliminarAsistencia(UUID asistenciaId) {
        if (!asistenciaRepository.existsById(asistenciaId)) {
            throw new ResponseStatusException(HttpStatus.NOT_FOUND, "Registro no encontrado");
        }
        asistenciaRepository.deleteById(asistenciaId);
    }

    public EstadisticasAsistenciaDTO getEstadisticasAlumno(UUID alumnoId, UUID cursoId) {
        Usuario alumno = usuarioRepository.findById(alumnoId)
                .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND, "Alumno no encontrado"));

        long total = asistenciaRepository.countByAlumnoAndCurso(alumnoId, cursoId);
        long presentes = asistenciaRepository.countByAlumnoAndCursoAndEstado(alumnoId, cursoId, EstadoAsistencia.PRESENTE);
        long ausentes = asistenciaRepository.countByAlumnoAndCursoAndEstado(alumnoId, cursoId, EstadoAsistencia.AUSENTE);
        long retrasos = asistenciaRepository.countByAlumnoAndCursoAndEstado(alumnoId, cursoId, EstadoAsistencia.RETRASO);
        long justificados = asistenciaRepository.countByAlumnoAndCursoAndEstado(alumnoId, cursoId, EstadoAsistencia.JUSTIFICADO);

        double porcentaje = total > 0 ? ((presentes + retrasos + justificados) * 100.0 / total) : 0;

        return new EstadisticasAsistenciaDTO(
                alumnoId,
                alumno.getNombre(),
                total,
                presentes,
                ausentes,
                retrasos,
                justificados,
                Math.round(porcentaje * 100.0) / 100.0
        );
    }

    public List<LocalDate> getFechasConAsistencia(UUID cursoId) {
        return asistenciaRepository.findDistinctFechasByCursoId(cursoId);
    }
}

File: osiconnect/src/services/asistenciaService.ts
import { api } from "@/services/api";

export type EstadoAsistencia = "PRESENTE" | "AUSENTE" | "RETRASO" | "JUSTIFICADO";

export interface AsistenciaDTO {
    id: string;
    alumnoId: string;
    alumnoNombre: string;
    cursoId: string;
    fecha: string;
    estado: EstadoAsistencia;
    observaciones: string | null;
}

export interface RegistroAsistenciaDTO {
    alumnoId: string;
    fecha: string;
    estado: EstadoAsistencia;
    observaciones?: string;
}

export interface EstadisticasAsistenciaDTO {
    alumnoId: string;
    alumnoNombre: string;
    totalClases: number;
    presentes: number;
    ausentes: number;
    retrasos: number;
    justificados: number;
    porcentajeAsistencia: number;
}

// Profesor/Admin: obtener toda la asistencia del curso
export async function getAsistenciaCurso(cursoId: string): Promise<AsistenciaDTO[]> {
    const res = await api.get(`/curso/${cursoId}/asistencia`);
    return res.data;
}

// Profesor/Admin: obtener asistencia por fecha
export async function getAsistenciaPorFecha(cursoId: string, fecha: string): Promise<AsistenciaDTO[]> {
    const res = await api.get(`/curso/${cursoId}/asistencia/fecha/${fecha}`);
    return res.data;
}

// Profesor/Admin: obtener fechas con asistencia registrada
export async function getFechasConAsistencia(cursoId: string): Promise<string[]> {
    const res = await api.get(`/curso/${cursoId}/asistencia/fechas`);
    return res.data;
}

// Alumno: obtener mi asistencia
export async function getMiAsistencia(cursoId: string): Promise<AsistenciaDTO[]> {
    const res = await api.get(`/curso/${cursoId}/asistencia/mi-asistencia`);
    return res.data;
}

// Alumno: obtener mis estadísticas
export async function getMisEstadisticas(cursoId: string): Promise<EstadisticasAsistenciaDTO> {
    const res = await api.get(`/curso/${cursoId}/asistencia/mi-asistencia/estadisticas`);
    return res.data;
}

// Profesor/Admin: obtener asistencia de un alumno
export async function getAsistenciaAlumno(cursoId: string, alumnoId: string): Promise<AsistenciaDTO[]> {
    const res = await api.get(`/curso/${cursoId}/asistencia/alumno/${alumnoId}`);
    return res.data;
}

// Profesor/Admin: obtener estadísticas de un alumno
export async function getEstadisticasAlumno(cursoId: string, alumnoId: string): Promise<EstadisticasAsistenciaDTO> {
    const res = await api.get(`/curso/${cursoId}/asistencia/alumno/${alumnoId}/estadisticas`);
    return res.data;
}

// Profesor/Admin: registrar asistencia individual
export async function registrarAsistencia(cursoId: string, registro: RegistroAsistenciaDTO): Promise<AsistenciaDTO> {
    const res = await api.post(`/curso/${cursoId}/asistencia`, registro);
    return res.data;
}

// Profesor/Admin: registrar asistencia múltiple
export async function registrarAsistenciaMultiple(cursoId: string, registros: RegistroAsistenciaDTO[]): Promise<void> {
    await api.post(`/curso/${cursoId}/asistencia/multiple`, registros);
}

// Profesor/Admin: actualizar asistencia
export async function actualizarAsistencia(
    cursoId: string,
    asistenciaId: string,
    estado: EstadoAsistencia,
    observaciones?: string
): Promise<AsistenciaDTO> {
    const res = await api.put(`/curso/${cursoId}/asistencia/${asistenciaId}`, {
        estado,
        observaciones: observaciones || null
    });
    return res.data;
}

// Profesor/Admin: eliminar registro de asistencia
export async function eliminarAsistencia(cursoId: string, asistenciaId: string): Promise<void> {
    await api.delete(`/curso/${cursoId}/asistencia/${asistenciaId}`);
}
